nucliadb/nucliadb/common/cluster/discovery/k8s.py

```diff
@@ -0,0 +1,192 @@
+        stub = nodewriter_pb2_grpc.NodeWriterStub(channel)
+        return await stub.GetMetadata(EmptyQuery())  # type: ignore
+
+    async def get_node_metadata(self, sts_name: str, force=False) -> IndexNodeMetadata:
+        if force or sts_name not in self.node_id_cache:
+            # hard coded assuming that nucliadb is used with current helm charts
+            address = f"{sts_name}.node.{self.settings.cluster_discovery_kubernetes_namespace}.svc.cluster.local"
+            metadata = await self._query_node_metadata(address)
+            self.node_id_cache[sts_name] = IndexNodeMetadata(
+                node_id=metadata.node_id,
+                name=sts_name,
+                shard_count=metadata.shard_count,
+                address=address,
+            )
+        return self.node_id_cache[sts_name]
+
+    async def update_node(self, event: EventType) -> None:
+        """
+        Update node metadata when a pod is updated.
+
+        This method will update global node state by utilizing the cluster manager
+        to add or remove nodes.
+        """
+        status: kubernetes_asyncio.client.models.v1_pod_status.V1PodStatus = event[
+            "object"
+        ].status
+        event_metadata: kubernetes_asyncio.client.models.v1_object_meta.V1ObjectMeta = (
+            event["object"].metadata
+        )
+
+        ready = status.container_statuses is not None
+        if event["type"] == "DELETED":
+            ready = False
+        elif status.container_statuses is not None:
+            container_statuses: list[
+                kubernetes_asyncio.client.models.v1_container_status.V1ContainerStatus
+            ] = status.container_statuses
+            for container_status in container_statuses:
+                if container_status.name not in ("reader", "writer", "sidecar"):
+                    continue
+                if not container_status.ready:
+                    ready = False
+                    break
+
+        sts_name = event_metadata.name
+        node_data = await self.get_node_metadata(sts_name)
+        if ready:
+            node = manager.get_index_node(node_data.node_id)
+            if node is None:
+                logger.debug(f"Adding node", extra={"node_id": node_data.node_id})
+                manager.add_index_node(
+                    IndexNode(
+                        id=node_data.node_id,
+                        address=node_data.address,
+                        shard_count=node_data.shard_count,
+                    )
+                )
+            else:
+                logger.debug(
+                    "Update node",
+                    extra={"sts_name": sts_name, "node_id": node_data.node_id},
+                )
+                node.address = node_data.address
+                node.shard_count = node_data.shard_count
+        else:
+            logger.debug(f"Remove node", extra={"node_id": node_data.node_id})
```

Comment:
What do you think about WARNING log level here? It should not be as chatty, and we may want to be able confirm when a node is actually down

Reply:
Fine with me!